### PR TITLE
Break out nested FileStream

### PR DESCRIPTION
**Pull request by JeremyKuhne**
This breaks out the nested FileStream in preparation for
moving the code down to corlib.

Will follow up with:

1. Rename of files (from Unix/Win32FS to FS.os.cs)
2. Moving of common code to FileStream.cs
3. Copy to corlib
4. Move ref to System.Runtime and type forward to corlib
5. Further prep work for bringing up UAP 10.1

@ianhays, @ericstj, @stephentoub, @jkotas 
#12633

### CODE REVIEW

**Comment by stephentoub** [comment]
@JeremyKuhne, can you summarize what was done here to make it easier to follow the changes?

**Comment by JeremyKuhne** [comment]
What I essentially did was "simply" change `Win32/UnixFileStream` to `FileStream` and remove the `parent` stream.

The shared `FileStream` is the common code (param validation, etc.). `FileStreamBase` is the abstract public surface.

I had to remove the `readonly` from members in the Unix impl. I want to try and move some or all of them back to `readonly` but I didn't want to make this change any harder to follow. For now the final initialization is not in a constructor, but in `Init()` methods.

I'll rename `Win32/UnixFileStream` to `FileStream.Win32/Unix` in the next change, didn't want to mess with it in this change as there were already files with these names that I deleted.

**Comment by stephentoub** [comment]
> What I essentially did

Thanks for the explanation.

> FileStreamBase is the abstract public surface.

FileStreamBase should not be the base type for FileStream and should not be exposed publicly.

**Comment by JeremyKuhne** [comment]
> FileStreamBase should not be the base type for FileStream

Why not? It already was. This allows at least some of the long term test scenarios people are asking for.

**Comment by stephentoub** [comment]
> It already was

No, it was the base for the inner stream that wasn't exposed in the public surface area.

> Why not?

Because it changes the public surface area.

**Comment by jkotas** [comment]
> FileStreamBase should not be the base type for FileStream

I agree as well.

**Comment by JeremyKuhne** [comment]
> Because it changes the public surface area.

I'll remove it, but how does inserting a class into the heirarchy matter?

**Comment by stephentoub** [comment]
> but how does inserting a class into the heirarchy matter?

It's not a breaking change, but it is new public surface area.  If we want to discuss adding that, going through the proper API reviews with justification for why we need it, etc., that's fine.  But it should be as an entirely separate change after this set of moving FileStream to corelib is complete.

**Comment by jkotas** [comment]
BTW: The FileStream class should be a wrapper around OS file handle (ie how it always used to be before UWP came online). The UWP case - where it wraps two different APIs - should be handled as outlier, the complexity from that should not bleed into what we do for regular platforms.

**Comment by JeremyKuhne** [comment]
Removed `FileStreamBase`

**Comment by JeremyKuhne** [comment]
This addresses #11707

**Comment by JeremyKuhne** [comment]
Another note: I'm planning to move the ArrayPool as well #12810.

**Comment by stephentoub** [comment]
@dotnet-bot test this please (Jenkins rebooted)

**Comment by ericstj** [comment]
> BTW: The FileStream class should be a wrapper around OS file handle (ie how it always used to be before UWP came online). The UWP case - where it wraps two different APIs - should be handled as outlier, the complexity from that should not bleed into what we do for regular platforms.

@jkotas that's going to be a challenge.  In the past we shared the same corlib between UWP and .NET Core.  UWP will have to meet the same requirements of .NETStandard2.0 and need to expose a real FileStream from corlib.  We'll either need to bring that complexity back into the single build of corlib, or start cross-compiling corlib for UWP.  @JeremyKuhne mentioned he'd open an issue to discuss this.

**Comment by jkotas** [comment]
> We'll either need to bring that complexity back into the single build of corlib, or start cross-compiling corlib for UWP

I do not think compiling two different CoreCLR corelib for Win32/UWP is worth it. (It is different for CoreRT - we are compiling different corelib for Win32/UWP there already.)

I would be fine if the damage is contained to a single line under ifdef to redirect the implementation for WinRT if necessary. Something like:

```
        public override void Flush()
        {
#if ENABLE_WINRT
            if (_override != null) { _override.Flush(); return; }
#endif

            ... regular Win32 Flush implementation ...
       }
```

**Comment by ericstj** [comment]
Yeah, that's similar to what @JeremyKuhne and I were discussing.   That override could be the WinRT PAL. The setting of the override would need to be controlled by some flag that only happened when running in UWP, and then only when we probe the path and decide it needs to use the WinRT PAL.

There's still a whole other problem of moving the WinRT PAL down into corlib.  Currently that depends on things like System.Runtime.WindowsRuntime.

**Comment by jkotas** [comment]
> There's still a whole other problem of moving the WinRT PAL down into corlib.

The implementation of WinRT PAL should stay out of corelib because of the dependencies. It should be initialized via reflection (or some other scheme) when running on UWP. There is prior art around this in CoreRT, like https://github.com/dotnet/corert/blob/6ef0d902bf388343c6f11d4622fc3402b9bcb6e0/src/System.Private.CoreLib/src/Internal/Runtime/Augments/WinRTInterop.cs#L102. E.g. there can be another method on this interface `public abstract Stream TryCreateWinRTStreamOverride(......);`.

**Comment by ericstj** [comment]
I guess we just need to be "really careful" not to have cycles, eh?  I suppose we could define an interface for the PAL in corlib so we at least only need to reflect for construction.

**Comment by jkotas** [comment]
> I guess we just need to be "really careful" not to have cycles, eh

Agree. 

> I suppose we could define an interface for the PAL in corlib

Would it make sense to just use the Stream class as the interface (beyond the construction)? There may be to some duplicated code compared to custom callback ... but it makes the contract between the components clear and stable.

**Comment by ericstj** [comment]
I think we need to delegate some non-stream members to the PAL, but could be wrong.  We also need to call back into filestream for some virtual members to get the overridden member when filestream is overridden.  Anyhow, those are details for future.
